compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -1620,6 +1633,47 @@ fn confirm_generator_candidate<'cx, 'tcx>(
         .with_addl_obligations(obligations)
 }
 
+fn confirm_as_repr_candidate<'cx, 'tcx>(
+    selcx: &mut SelectionContext<'cx, 'tcx>,
+    obligation: &ProjectionTyObligation<'tcx>,
+) -> Progress<'tcx> {
+    let tcx = selcx.tcx();
+
+    let self_ty = selcx.infcx().shallow_resolve(obligation.predicate.self_ty());
+    // We get here from `poly_project_and_unify_type` which replaces bound vars
+    // with placeholders
+    debug_assert!(!self_ty.has_escaping_bound_vars());
+    let substs = tcx.mk_substs([self_ty.into()].iter());
+
+    let as_repr_repr_def_id = tcx.require_lang_item(LangItem::AsReprRepr, None);
+
+    let adt_def = match self_ty.ty_adt_def() {
+        Some(adt_def) => adt_def,
+        None => bug!("Cannot project as_repr shim for {:?} which is not an ADT", self_ty),
+    };
+
+    let disciminant_type = if Some(adt_def.did) == tcx.lang_items().discriminant_struct() {
+        if let ty::Adt(_, substs) = self_ty.kind() {
+            substs.type_at(0).discriminant_ty(tcx)
+        } else {
+            bug!(
+                "core::mem::Discriminant is known to be a struct, but found it not to be an ADT when projecting AsRepr candidate"
```

Comment:
[nitpick] split over two lines with a ` \`. `x.py fmt` won't try to do so inside macros, but it is nice to keep things consistent.